include/battle.h

```diff
@@ -1243,4 +1322,145 @@ static inline bool32 IsBattlerInvalidForSpreadMove(u32 battlerAtk, u32 battlerDe
         || (battlerDef == BATTLE_PARTNER(battlerAtk) && (moveTarget == MOVE_TARGET_BOTH));
 }
 
+static inline u32 MaxRaiseOrLowerStatAmount(u32 battler, u32 stat, bool32 lowering)
+{
+    return lowering ? (gBattleMons[battler].statStages[stat] - MIN_STAT_STAGE) : (MAX_STAT_STAGE - gBattleMons[battler].statStages[stat]);
+}
+
+static inline union StatChanger PrepareStatChangerAny(union StatFlags stats, s32 stage, bool32 doNotSetStatId)
+{
+    return (union StatChanger) ((((union StatChanger) {
+        .attack = stats.attack, // All 1 or 0 depending on if the stat is selected in 'stats'
+        .defense = stats.defense,
+        .speed = stats.speed,
+        .spAttack = stats.spAttack,
+        .spDefense = stats.spDefense,
+        .accuracy = stats.accuracy,
+        .evasion = stats.evasion,
+    }).value
+        * abs(stage)) // Multiplies by the stage
+        | (stage < 0) // isNegative
+        | ((doNotSetStatId || (popcount(stats.allStats) > 1)) ? 0 : (log2Int(stats.allStats) << 1))); // statId;
+}
+
+static inline union StatChanger CalcStatChangerValue(u32 statId, s32 stage)
+{
+    return PrepareStatChangerAny(1 << statId, stage, FALSE);
+}
+
+static inline void SetStatChanger(u32 statId, s32 stage)
+{
+    gBattleScripting.statChanger = CalcStatChangerValue(statId, stage);
+}
+
+static inline union StatChanger OrStatChangers(union StatChanger statChangerA, union StatChanger statChangerB)
+{
+    return (union StatChanger) (statChangerA.value | statChangerB.value);
+}
```

Comment:
This seems incredibly sketchy, I would say that none of the fields in a stat changer are appropriate to bitwise or together. I think this would only make sense if you know that A and B have mutually exclusive members, but if you know that then presumably you know which members are set and instead we could have some function(s) whose name(s) represents the members that are being set.

Reply:
Can I ask what solution you would suggest?

Reply:
afaict `OrStatChangers` is only used in `ChangeStatBuffs` and `Cmd_statbuffchange` and `statChangerB` is always `PrepareStatChangerAny(stats, 1, TRUE)`. So my first thought would be to just write a function that applies the appropriate bits to a `statChanger` argument.

With `|`, if `statChangerA.attack = 1` and `statChangerB.attack = 2` then `returned.attack = 3` which seems like it can't possibly be the intended result. It probably makes more sense to either override the stat amounts or to only update zero stat amounts (giving the result `2` or `1` respectively). I suspect it can't make sense to combine `statChangerA.isNegative != statChangerB.isNegative`, and I suspect that for `statChangerA.statId != statChangerB.statId` we should set to `STAT_MULTIPLE`?

Reply:
Added a better solution